x-pack/legacy/plugins/maps/public/layers/util/es_agg_utils.ts

```diff
@@ -0,0 +1,51 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+import { i18n } from '@kbn/i18n';
+import _ from 'lodash';
+import { IndexPattern, IFieldType } from '../../../../../../../src/plugins/data/public';
+
+export function getField(indexPattern: IndexPattern, fieldName: string) {
+  const field = indexPattern.fields.getByName(fieldName);
+  if (!field) {
+    throw new Error(
+      i18n.translate('xpack.maps.source.esSearch.fieldNotFoundMsg', {
+        defaultMessage: `Unable to find '{fieldName}' in index-pattern '{indexPatternTitle}'.`,
+        values: { fieldName, indexPatternTitle: indexPattern.title },
+      })
+    );
+  }
+  return field;
+}
+
+export function addFieldToDSL(dsl: any, field: IFieldType) {
```

Comment:
imho, we know a little bit more about these objects to have to use `any`. They're basically object-literals. I think a Record would be appropriate here.

e.g.

We could use generics for the value:

```
export function addFieldToDSL<T>(dsl: Record<string | number, T>, field: IFieldType) {
  return !field.scripted
    ? { ...dsl, field: field.name }
    : {
        ...dsl,
        script: {
          source: field.script,
          lang: field.lang,
        },
      };
}

export function extractPropertiesFromBucket<T>(
  bucket: Record<string | number, T>,
  ignoreKeys: string[] = []
) {
  const properties: Record<string | number, T> = {};
  for (const key in bucket) {
    if (bucket.hasOwnProperty(key)) {
      if (ignoreKeys.includes(key)) {
        continue;
      }

      const bucketValue: T = bucket[key];
      if (_.has(bucketValue, 'value')) {
        properties[key] = _.get(bucketValue, 'value');
      } else if (_.has(bucketValue, 'buckets')) {
        properties[key] = _.get(bucketValue, 'buckets[0].key');
      } else {
        properties[key] = bucketValue;
      }
    }
  }
  return properties;
}
```
But `unknown` would work too imho.

```
export function addFieldToDSL<T>(dsl: Record<string | number, T>, field: IFieldType) {
  return !field.scripted
    ? { ...dsl, field: field.name }
    : {
        ...dsl,
        script: {
          source: field.script,
          lang: field.lang,
        },
      };
}

interface IValued {
  value: unknown
}

export function extractPropertiesFromBucket(
  bucket: Record<string | number, unknown>,
  ignoreKeys: string[] = []
) {
  const properties: Record<string | number, unknown> = {};
  for (const key in bucket) {
    if (bucket.hasOwnProperty(key)) {
      if (ignoreKeys.includes(key)) {
        continue;
      }

      const bucketValue = bucket[key] as object;
      if (_.has(bucketValue, 'value')) {
        const valued = bucketValue as IValued;
        properties[key] = valued.value;
      } else if (_.has(bucketValue, 'buckets')) {
        properties[key] = _.get(bucketValue, 'buckets[0].key');
      } else {
        properties[key] = bucketValue;
      }
    }
  }
  return properties;
}
```

Reply:
For `addFieldToDSL`, limiting the DSL to string and number is not a valid assumption. Elasticsearch aggregation DSL can also have objects. For example, https://github.com/elastic/kibana/issues/56876 will add `bounds` object to the DSL. `any` is an accurate type descriptor for the `dsl` parameter. The method just adds either `field` or `script` (which is an object) to `dsl`. The shape of the original `dsl` is of no concern to the method and does not effect the output in any way. I think this is a strong case for specifying `dsl` as any.